src/atomate2/vasp/jobs/defect.py

```diff
@@ -3,27 +3,324 @@
+
+    for defect in defects:
+        q_job = run_all_charge_states(
+            defect,
+            sc_mat=sc_mat,
+            relax_maker=relax_maker,
+            defect_index=f"{name_counter[defect.name]}",
+        )
+        defect_q_jobs.append(q_job)
+        output[f"{defect.name}-{name_counter[defect.name]}"] = dict(
+            defect=defect,
+            results=q_job.output,
+        )
+        name_counter[defect.name] += 1
+    return Response(output=output, replace=defect_q_jobs)
+
+
+@job
+def run_all_charge_states(
+    defect: Defect,
+    relax_maker: RelaxMaker,
+    sc_mat: NDArray | None = None,
+    defect_index: str = "",
+    add_info: dict | None = None,
+) -> Response:
+    """Perform charge defect supercell calculations and save the Hartree potential.
+
+    Run a ISIF2 calculation for each available charge state of the defect.
+    Ensure that the LOCPOT file is stored in the output.
+
+    Parameters
+    ----------
+    defect:
+        A defect object representing the defect in a unit cell.
+    relax_maker:
+        A RelaxMaker object to use for the atomic relaxation.
+    sc_mat:
+        The supercell matrix. If None, the code will attempt to create a
+        nearly-cubic supercell.
+    defect_index:
+        Additional index to give unique names to the defect calculations.
+    add_info:
+        Additional information to store with the defect cell relaxation calculation.
+        By default only the defect object and charge state are stored.
+
+    Returns
+    -------
+    Response
+        A response object containing the summary of the calculations for different
+        charge states.
+    """
+    jobs = []
+    outputs = dict()
+    sc_def_struct = defect.get_supercell_structure(sc_mat=sc_mat)
+    for qq in defect.get_charge_states():
+        suffix = (
+            f" {defect.name} q={qq}"
+            if defect_index == ""
+            else f" {defect.name}-{defect_index} q={qq}"
+        )
+        charged_struct = sc_def_struct.copy()
+        charged_struct.set_charge(qq)
+        charged_relax = relax_maker.make(charged_struct)
+        charged_relax.append_name(suffix)
+
+        # write some provenances data in info.json file
+        info = {
+            "defect": defect,
+            "charge_state": qq,
+        }
+        if add_info is not None:
+            info.update(add_info)
+        charged_relax.update_maker_kwargs(
+            {"_set": {"write_additional_data->info:json": info}}, dict_mod=True
+        )
+        jobs.append(charged_relax)
+        charged_output: TaskDocument = charged_relax.output
+
+        outputs[qq] = {
+            "structure": charged_output.structure,
+            "entry": charged_output.entry,
+            "dir_name": charged_output.dir_name,
+            "uuid": charged_relax.uuid,
+        }
+    add_flow = Flow(jobs, output=outputs)
+    return Response(output=outputs, replace=add_flow)
+
+
+@job
+def collect_defect_outputs(
```

Comment:
This returns quite a complicated results object. Can you:
1. Create a new schema for it.
2. Add a `from_objects` or something class method that generates schema from the calculation data. This should rely only on the pymatgen objects themselves rather than the calculation directories (although it can include the calculation directories as inputs so they get stored in the schema as metadata). This function can then be used by a future builder.
3. Add another classmethod that accepts a list of directories etc, that loads the files and calls the classmethod from point 2.
4. In this job, just call the classmethod from point 3 and return the schema.

E.g., this should follow how we do it in the ElectronPhonon workflow (see [here](https://github.com/materialsproject/atomate2/blob/0fe95429145d7b2a0988bc4e3a36b924516f9711/src/atomate2/vasp/flows/elph.py#L167-L189) and [here](https://github.com/materialsproject/atomate2/blob/0fe95429145d7b2a0988bc4e3a36b924516f9711/src/atomate2/vasp/jobs/elph.py#L194)) and Phonopy workflow (see [here](https://github.com/materialsproject/atomate2/blob/0fe95429145d7b2a0988bc4e3a36b924516f9711/src/atomate2/vasp/flows/phonons.py#L314-L336) and [here](https://github.com/materialsproject/atomate2/blob/0fe95429145d7b2a0988bc4e3a36b924516f9711/src/atomate2/vasp/jobs/phonons.py#L193)).

For these classmethods, it is much better to be explicit with what arguments are required rather than relying on a dictionary with the data which is much harder to document and reproduce. When we eventually add a builder/create the CP2K defect workflow, this will reduce the code duplication significantly.